File: Quick_Sort.py
# A list to sort
arr = [6,10,13,5,8,3,2,11];

#It Sorts and Finds Divide Location
def divide(the_list,start,end):
    
    #First element is used as pivot 
    pivot=the_list[start]
    print("pivot = ",pivot)
    loc = start

    #Find lower elements than pivot
    for i in range(start+1,end):
        if (the_list[i]<pivot):
            loc = loc + 1
            tmp = the_list[loc]
            the_list[loc] = the_list[i]
            the_list[i] = tmp
    
    #Locate the pivot        
    tmp = the_list[loc]
    the_list[loc] = the_list[start]
    the_list[start] = tmp
    print("The list after sort at this stage",the_list)
    return loc

#Recursive Quick Sort
def quick_sort(a_list,st,en):
    if st<en:
        location = divide(a_list,st,en)
        quick_sort(a_list,st,location)
        quick_sort(a_list,location+1,en)

#Client Query
print("The list before divide",arr)
quick_sort(arr,0,8)

File: Quick_Sort.cpp
#include <iostream>
using namespace std;

int divide_array(int *, int);
void Quick_Sort(int list[],int s, int e);

int main() 
{
    int arr[] = {6,10,13,5,8,3,2,11};
    cout << "Hello, World! \n";
    //divide_array(arr, 8);
    Quick_Sort(arr,0,7);
    cout<<"\narr= ";
    for (int i = 0; i < 7; i++)
        {
            cout<<"\t"<< arr[i];

        }
    return 0;
}


int divide_array(int *ar,int bas, int len){
    
    int pivot;
    pivot = ar[bas];
    cout<<"\nPIVOT ELEMENT : "<<pivot;
    cout<<"\n bas = "<<bas<<" son = " << len <<"\n";
    cout<<"\n The actual list : \n";
    
    for (int i = bas; i < len; i++)
    {
        cout<<ar[i]<<"\t";
    }
    
    int p_loc = bas;

    int p = 0;

    for (int i = bas+1; i < len; i++)
    {
        if(ar[i]<pivot){
           
            p_loc += 1;
            int tmp;
            tmp = ar[i];
            ar[i] = ar[p_loc];
            ar[p_loc] = tmp;
        }

        p = ar[p_loc];
    }

 

    int tmp;
    tmp = ar[bas];
    ar[bas] = ar[p_loc];
    ar[p_loc] = tmp;

    

    cout<<"\n Sorted list in this stage "<<"\n";
        for (int i = 0; i < len; i++)
        {
            cout<<ar[i]<<"\t";
        }

        return p_loc;
}


void Quick_Sort(int *list,int s, int e){
    int loc;
    if (s<e)
    {
        loc = divide_array(list,s,e);
        cout<<"\nloc = "<<loc;
        Quick_Sort(list,s,loc-1);
        Quick_Sort(list,loc+1,e);
    }
    
 
}